File: Python/Dynamic Programming/5. Longest Palindromic Substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        if n == 1: return s
        if n == 2: return s if s[0]==s[1] else s[0]
        # row 0 for even length, row 1 for odd length
        memo = [[0 for _ in range(n)], [1 for _ in range(n)]]
        subStrLen = 1
        endIdx = 0
        for i in range(1, n):
            if s[i]==s[i-1]:
                memo[0][i] = 1
                subStrLen = 2
                endIdx = i
        # substring length starting from 3 up to n
        for i in range(3, n+1):
            row = i%2
            # end index starting from n-1 down to i-1
            for j in range(n-1, i-2, -1):
                memo[row][j] = 0
                if memo[row][j-1] == 1 and s[j]==s[j-i+1]:
                    memo[row][j] = 1
                    subStrLen = i
                    endIdx = j
        return s[endIdx-subStrLen+1:endIdx+1]
        

File: C++/Dynamic Programming/5. Longest Palindromic Substring.cpp
// This DP solution is O(n^2) in time, O(n) in space.
// Interesting solution: Manacher's algorithm, O(n) in time.

class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.length();
        if (n == 1) return s;
        if (n == 2) return s[0]==s[1] ? s : string(1, s[0]);
        int memo[2][n];
        int subStrLen = 1, endIdx = 0;
        // One character
        for (int i = 0; i < n; i++) memo[1][i] = 1;
        // Two characters
        for (int i = 1; i < n; i++) {
            if (s[i-1] == s[i]) {
                subStrLen = 2;
                endIdx = i;
                memo[0][i] = 1;
            }
            else {
                memo[0][i] = 0;
            }
        }
        // Starting from 3 characters up to n characters
        for (int i = 3; i <= n; i++) {
            // row 0 if even, row 1 if odd
            int row = i%2;
            // Index of character in s: from n-1 down to i-1
            for (int j = n-1; j >= i-1; j--) {
                memo[row][j] = 0;
                if (memo[row][j-1] == 1 && s[j] == s[j-i+1]) {
                    memo[row][j] = 1;
                    subStrLen = i;
                    endIdx = j;
                }
            }
        }
        return s.substr(endIdx-subStrLen+1, subStrLen);
    }
};